### ISSUE TITLE
SuperLU solver not freeing memory

### ISSUE THREAD

**Issue by cdelv** [mfem/mfem#2413]
I have set up the solution of a system of equations with superlu. The code is the following
```c++
    //Create the complete bilinear operator:
    //
    //   H = [ M    C ] 
    //       [ C^t  cm D ] 
    Array2D<HypreParMatrix*> hBlocks(2,2);
    hBlocks = NULL;
    hBlocks(0, 0) = M;
    hBlocks(0, 1) = C;
    hBlocks(1, 0) = C->Transpose();
    hBlocks(1, 1) = D;

    Array2D<double> blockCoeff(2,2);
    blockCoeff(0, 0) = 1.;
    blockCoeff(0, 1) = -1.;
    blockCoeff(1, 0) = -1.;
    blockCoeff(1, 1) = -1.;

    HypreParMatrix *H = HypreParMatrixFromBlocks(hBlocks, &blockCoeff);

    SuperLUSolver *superlu = new SuperLUSolver(MPI_COMM_WORLD);
    Operator *SLU_A = new SuperLURowLocMatrix(*H);
    superlu->SetOperator(*SLU_A);
    superlu->SetPrintStatistics(true);
    superlu->SetSymmetricPattern(true);
    superlu->SetColumnPermutation(superlu::PARMETIS);
    superlu->SetIterativeRefine(superlu::SLU_DOUBLE);

    //Solve the linear system Ax=B
    X.Randomize();
    superlu->Mult(B, X);

    //Recover the solution on each proccesor
    w->Distribute(&(X.GetBlock(0)));
    for (int ii = 0; ii < w->Size(); ii++)
        (*w)(ii) += (*w_aux)(ii); 
    
    psi->Distribute(&(X.GetBlock(1)));
    for (int ii = 0; ii < psi->Size(); ii++)
        (*psi)(ii) += (*psi_aux)(ii); 

    v = new ParGridFunction(fespace_v);
    GradientGridFunctionCoefficient psi_grad(psi);
    MatrixFunctionCoefficient rot(dim, rot_f);
    MatrixVectorProductCoefficient rV(rot, psi_grad);
    v->ProjectDiscCoefficient(rV, GridFunction::ARITHMETIC);

    //Delete used memory
    delete H;
    delete superlu;
    delete SLU_A;
```
Then I used the same code to solve for the velocity in a time-dependent advection-diffusion equation. I noticed that with every iteration the program kept increasing its RAM usage. So, I went back to my superlu static solver and put it in a for loop and found out that it behaved in the same way. I really don't understand because delete superlu should call the destructor. Further on the program, all the matrixes are deleted and at the end it frees all the memory. The problem is that it doesn't free the memory before the program ends.

Can someone help me solve this problem, or give some advice on how to find the thing that's not freeing memory? I used sanitizers and valgrind but they dint say much.

**Comment by cdelv** [comment]
To complement this, I have been searching for a potential solution and found this http://www.elmerfem.org/forum/viewtopic.php?t=3625.  I also have found people reporting similar leaks.

**Comment by psocratis** [comment]
@cdelv, Thanks for reporting this. 

Just for us to test this can you let us know if you are using mfem's master branch or not and with which version of SuperLU you are linking with? 

Thanks, 

Socratis

**Comment by psocratis** [comment]
@cdelv, would you mind trying `mfem/examples/superlu/ex1p.cpp` , putting the superlu solver in a loop and let us know if you see the same kind of issue? 

Thanks, 

Socratis

**Comment by cdelv** [comment]
Ok, I will and let you know.

**Comment by cdelv** [comment]
@psocratis I ran example 1 and had similar behavior. However, if the for loop only gets the mult function there's no leak. If I loop over all the solver set up it happens, this is step 13. To make the memory usage big enough to have a significant leak I used 2 parallel refinements with 4 cores. The rest of the parameters I left as default in the example.

**Comment by psocratis** [comment]
@cdelv, hmm... thanks for doing that. I did the same test as well and I don't see a leak. I do use MFEM master branch though along with SuperLU_dist v.6.3.1. I can look into it more carefully, and get back to you. 

Thanks, 

Socratis

**Comment by cdelv** [comment]
@psocratis don't you get that each time it iterates it uses more ram? I found that Valgrind doesn't report it as a leak I guess because it frees all the memory at the end, but it doesn't free it when I call delete.

**Comment by psocratis** [comment]
> @psocratis don't you get that each time it iterates it uses more ram? I found that Valgrind doesn't report it as a leak I guess because it frees all the memory at the end, but it doesn't free it when I call delete.

Exactly, I don't see the ram increasing ...

**Comment by cdelv** [comment]
then it must be a bug in the spack version of MFEM or SUPERLU. How fixes that? Should I share this with spack developers?
Also, I'm using openmpi, I don't have mpich.

**Comment by psocratis** [comment]
I also just noticed this #2368 . So let me look into this more carefully, and I will get back to you. 

Thanks, 

Socratis

**Comment by cdelv** [comment]
@psocratis I checked and spack uses superlu_dist version 6.4.0 and mfem 4.2.0

**Comment by cdelv** [comment]
@psocratis I installed MFEM from the master branch of the repository and superlu_dist 6.3.1 and the memory problem was still there. I also used superlu_dist 6.4.0 with the same results. I'm using mpi 4.0.5.  
My Linux distro is Pop OS 20.10 based on ubuntu and I used the CMake install.

**Comment by cdelv** [comment]
Update, I change the code to use smart pointers, so the memory management is done automatically. That way is not an implementation error.
```
    std::unique_ptr<SuperLUSolver> superlu(new SuperLUSolver(MPI_COMM_WORLD));
    std::unique_ptr<SuperLURowLocMatrix> SLU_A(new SuperLURowLocMatrix(*H));
```
for example. The problem with the memory still happens. So the problem has to be in the destructor of the solver.

**Comment by psocratis** [comment]
> for example. The problem with the memory still happens. So the problem has to be in the destructor of the solver.

Just to make sure....you put a loop maybe 100-200 times and it always keeps increasing? How are checking the ram usage btw?

**Comment by psocratis** [comment]
> > for example. The problem with the memory still happens. So the problem has to be in the destructor of the solver.
> 
> Just to make sure....you put a loop maybe 100-200 times and it always keeps increasing? How are you checking the ram usage btw?

Ok I think I might be on to something .... 

Can you try puting this call `DismantleGrid();` in the end of the `SuperLUSolver` destructor and let me know if this goes away. I did see some leaks eventually, and with that addition, they went away ... 

Thanks

**Comment by psocratis** [comment]
So, eventually I was able to pin down a leak. 

Following the comment [here](https://github.com/mfem/mfem/blob/d0caa1c432b234aa614fde5df87ee5e1d89cb92d/linalg/superlu.hpp#L103-L104), I put the call `superlu->DismantleGrid();` right after the solve (`superlu->Mult(B, X);`) and before deleting the pointers. There are no leaks reported after doing this. Can you try this as well and let me know if anything changes? 

Thanks, 

Socratis

**Comment by cdelv** [comment]
@psocratis  I monitored the ram usage during execution with htop, a resource monitor. I know that thas not the most accurate way of doing it but for me, it has worked well many times before.  And for the tests, I shut down all the applications except for the terminal. I noticed the problem because the program filled the memory of my computer and crashed. My friends reported the same thing in their machines. 

Following your advice, I did the following to the end of the for of example 1
```
   superlu->SetOperator(*SLU_A);
   superlu->SetPrintStatistics(true);
   superlu->Mult(B, X);
   superlu ->DismantleGrid();

   delete superlu;
   delete SLU_A;
```
And didn't work. We also tried in other programs that use superlu solver calling it multiple times and dint work. Do you mean we should edit the MFEM source code?

**Comment by psocratis** [comment]
> And didn't work. We also tried in other programs that use superlu solver calling it multiple times and dint work. Do you mean we should edit the MFEM source code?

@cdelv 
Thanks for checking this. No I don't mean editing the MFEM source code, not just yet anyway. Can you try commenting out the `SetOperator` and the `mult` calls just to make sure that this is indeed coming from this? I'll have to dig into this more carefully. 

Thanks, 
Socratis

**Comment by cdelv** [comment]
@psocratis 
Well it ran very very fast, so I increase the number of iterations to 10^6. The program used little memory and maintained quite constant. So I guess that the problem is in the mult function after all. I did the test with both mult and set operator commented and then with only mult commented and it behaved well. 

If I put the for loop only in the mult function it doesn't leak. Its quite weird.

**Comment by jandrej** [comment]
Measuring memory usage of a program with `top` is not deterministic, nor an indicator that a program is leaking memory. Depending on the decision of the memory pool of the operating system, the memory usage won't decrease even if there is an instruction to release memory in your program.

Can you provide a _minimal_ working example where we should be able to reproduce the error?

**Comment by cdelv** [comment]
yeah sure, where can I send you the code?

**Comment by jandrej** [comment]
A minimal working example should be posted here in a code block.

**Comment by cdelv** [comment]
Well, I will post example 1 with the for loop. More complicated code wont fit in a simple code block. Also, variations like calling the solver from a function exhibit the same behavior. 

```
#include "mfem.hpp"
#include <fstream>
#include <iostream>

#ifndef MFEM_USE_SUPERLU
#error This example requires that MFEM is built with MFEM_USE_SUPERLU=YES
#endif

using namespace std;
using namespace mfem;

int main(int argc, char *argv[])
{
   // 1. Initialize MPI.
   int num_procs, myid;
   MPI_Init(&argc, &argv);
   MPI_Comm_size(MPI_COMM_WORLD, &num_procs);
   MPI_Comm_rank(MPI_COMM_WORLD, &myid);

   // 2. Parse command-line options.
   const char *mesh_file = "mesh.msh";
   int order = 1;
   const char *device_config = "cpu";
   bool visualization = true;
   int slu_colperm = 4;
   int slu_rowperm = 1;
   int slu_iterref = 2;

   OptionsParser args(argc, argv);
   args.AddOption(&mesh_file, "-m", "--mesh",
                  "Mesh file to use.");
   args.AddOption(&order, "-o", "--order",
                  "Finite element order (polynomial degree) or -1 for"
                  " isoparametric space.");
   args.AddOption(&device_config, "-d", "--device",
                  "Device configuration string, see Device::Configure().");
   args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
                  "--no-visualization",
                  "Enable or disable GLVis visualization.");
   args.AddOption(&slu_colperm, "-cp", "--colperm",
                  "SuperLU Column Permutation Method:  0-NATURAL, 1-MMD-ATA "
                  "2-MMD_AT_PLUS_A, 3-COLAMD, 4-METIS_AT_PLUS_A, 5-PARMETIS "
                  "6-ZOLTAN");
   args.AddOption(&slu_rowperm, "-rp", "--rowperm",
                  "SuperLU Row Permutation Method:  0-NOROWPERM, 1-LargeDiag");
   args.AddOption(&slu_iterref, "-rp", "--rowperm",
                  "SuperLU Iterative Refinement:  0-NOREFINE, 1-Single, "
                  "2-Double, 3-Extra");

   args.Parse();
   if (!args.Good())
   {
      if (myid == 0)
      {
         args.PrintUsage(cout);
      }
      MPI_Finalize();
      return 1;
   }
   if (myid == 0)
   {
      args.PrintOptions(cout);
   }

   // 3. Enable hardware devices such as GPUs, and programming models such as
   //    CUDA, OCCA, RAJA and OpenMP based on command line options.
   Device device(device_config);
   if (myid == 0) { device.Print(); }

   // 4. Read the (serial) mesh from the given mesh file on all processors.  We
   //    can handle triangular, quadrilateral, tetrahedral, hexahedral, surface
   //    and volume meshes with the same code.
   Mesh mesh(mesh_file, 1, 1);
   int dim = mesh.Dimension();

   // 5. Refine the serial mesh on all processors to increase the resolution. In
   //    this example we do 'ref_levels' of uniform refinement. We choose
   //    'ref_levels' to be the largest number that gives a final mesh with no
   //    more than 1,000 elements.
   {
      int ref_levels =
         (int)floor(log(1000./mesh.GetNE())/log(2.)/dim);
      for (int l = 0; l < ref_levels; l++)
      {
         mesh.UniformRefinement();
      }
   }

   // 6. Define a parallel mesh by a partitioning of the serial mesh. Refine
   //    this mesh further in parallel to increase the resolution. Once the
   //    parallel mesh is defined, the serial mesh can be deleted.
   ParMesh pmesh(MPI_COMM_WORLD, mesh);
   mesh.Clear();
   {
      int par_ref_levels = 2;
      for (int l = 0; l < par_ref_levels; l++)
      {
         pmesh.UniformRefinement();
      }
   }

   // 7. Define a parallel finite element space on the parallel mesh. Here we
   //    use continuous Lagrange finite elements of the specified order. If
   //    order < 1, we instead use an isoparametric/isogeometric space.
   FiniteElementCollection *fec;
   bool delete_fec;
   if (order > 0)
   {
      fec = new H1_FECollection(order, dim);
      delete_fec = true;
   }
   else if (pmesh.GetNodes())
   {
      fec = pmesh.GetNodes()->OwnFEC();
      delete_fec = false;
      if (myid == 0)
      {
         cout << "Using isoparametric FEs: " << fec->Name() << endl;
      }
   }
   else
   {
      fec = new H1_FECollection(order = 1, dim);
      delete_fec = true;
   }
   ParFiniteElementSpace fespace(&pmesh, fec);
   HYPRE_Int size = fespace.GlobalTrueVSize();
   if (myid == 0)
   {
      cout << "Number of finite element unknowns: " << size << endl;
   }

   // 8. Determine the list of true (i.e. parallel conforming) essential
   //    boundary dofs. In this example, the boundary conditions are defined
   //    by marking all the boundary attributes from the mesh as essential
   //    (Dirichlet) and converting them to a list of true dofs.
   Array<int> ess_tdof_list;
   if (pmesh.bdr_attributes.Size())
   {
      Array<int> ess_bdr(pmesh.bdr_attributes.Max());
      ess_bdr = 1;
      fespace.GetEssentialTrueDofs(ess_bdr, ess_tdof_list);
   }

   // 9. Set up the parallel linear form b(.) which corresponds to the
   //    right-hand side of the FEM linear system, which in this case is
   //    (1,phi_i) where phi_i are the basis functions in fespace.
   ParLinearForm b(&fespace);
   ConstantCoefficient one(1.0);
   b.AddDomainIntegrator(new DomainLFIntegrator(one));
   b.Assemble();

   // 10. Define the solution vector x as a parallel finite element grid function
   //     corresponding to fespace. Initialize x with initial guess of zero,
   //     which satisfies the boundary conditions.
   ParGridFunction x(&fespace);
   x = 0.0;

   // 11. Set up the parallel bilinear form a(.,.) on the finite element space
   //     corresponding to the Laplacian operator -Delta, by adding the Diffusion
   //     domain integrator.
   ParBilinearForm a(&fespace);
   a.AddDomainIntegrator(new DiffusionIntegrator(one));

   // 12. Assemble the parallel bilinear form and the corresponding linear
   //     system, applying any necessary transformations such as: parallel
   //     assembly, eliminating boundary conditions, applying conforming
   //     constraints for non-conforming AMR, static condensation, etc.
   a.Assemble();

   OperatorPtr A;
   Vector B, X;
   a.FormLinearSystem(ess_tdof_list, x, b, A, X, B);

   // 13. Solve the linear system A X = B utilizing SuperLU.
   for(int i=0; i<1000000; i++)
     {
       SuperLUSolver *superlu = new SuperLUSolver(MPI_COMM_WORLD);
       Operator *SLU_A = new SuperLURowLocMatrix(*A.As<HypreParMatrix>());
       superlu->SetPrintStatistics(true);
       superlu->SetSymmetricPattern(false);
       
       if (slu_colperm == 0)
	 {
	   superlu->SetColumnPermutation(superlu::NATURAL);
	 }
       else if (slu_colperm == 1)
	 {
	   superlu->SetColumnPermutation(superlu::MMD_ATA);
	 }
       else if (slu_colperm == 2)
	 {
	   superlu->SetColumnPermutation(superlu::MMD_AT_PLUS_A);
	 }
       else if (slu_colperm == 3)
	 {
	   superlu->SetColumnPermutation(superlu::COLAMD);
	 }
       else if (slu_colperm == 4)
	 {
	   superlu->SetColumnPermutation(superlu::METIS_AT_PLUS_A);
	 }
       else if (slu_colperm == 5)
	 {
	   superlu->SetColumnPermutation(superlu::PARMETIS);
	 }
       else if (slu_colperm == 6)
	 {
      superlu->SetColumnPermutation(superlu::ZOLTAN);
	 }
       
       if (slu_rowperm == 0)
	 {
      superlu->SetRowPermutation(superlu::NOROWPERM);
	 }
       else if (slu_rowperm == 1)
	 {
#ifdef MFEM_USE_SUPERLU5
	   superlu->SetRowPermutation(superlu::LargeDiag);
#else
	   superlu->SetRowPermutation(superlu::LargeDiag_MC64);
#endif
	 }
       
       if (slu_iterref == 0)
	 {
	   superlu->SetIterativeRefine(superlu::NOREFINE);
	 }
       else if (slu_iterref == 1)
	 {
	   superlu->SetIterativeRefine(superlu::SLU_SINGLE);
	 }
       else if (slu_iterref == 2)
	 {
	   superlu->SetIterativeRefine(superlu::SLU_DOUBLE);
	 }
       else if (slu_iterref == 3)
	 {
	   superlu->SetIterativeRefine(superlu::SLU_EXTRA);
	 }
       superlu->SetOperator(*SLU_A);
       superlu->SetPrintStatistics(true);
       superlu->Mult(B, X);
       superlu ->DismantleGrid();
       
       
       delete superlu;
       delete SLU_A;
     }

   // 14. Recover the parallel grid function corresponding to X. This is the
   //     local finite element solution on each processor.
   a.RecoverFEMSolution(X, b, x);

   // 15. Save the refined mesh and the solution in parallel. This output can
   //     be viewed later using GLVis: "glvis -np <np> -m mesh -g sol".
   {
      ostringstream mesh_name, sol_name;
      mesh_name << "mesh." << setfill('0') << setw(6) << myid;
      sol_name << "sol." << setfill('0') << setw(6) << myid;

      ofstream mesh_ofs(mesh_name.str().c_str());
      mesh_ofs.precision(8);
      pmesh.Print(mesh_ofs);

      ofstream sol_ofs(sol_name.str().c_str());
      sol_ofs.precision(8);
      x.Save(sol_ofs);
   }

   // 16. Send the solution by socket to a GLVis server.
   if (visualization)
   {
      char vishost[] = "localhost";
      int  visport   = 19916;
      socketstream sol_sock(vishost, visport);
      sol_sock << "parallel " << num_procs << " " << myid << "\n";
      sol_sock.precision(8);
      sol_sock << "solution\n" << pmesh << x << flush;
   }

   // 17. Free the used memory.
   if (delete_fec)
   {
      delete fec;
   }
   MPI_Finalize();

   return 0;
}
```

**Comment by jandrej** [comment]
And what mesh are you using? Does it happen with the default meshes in mfem?

**Comment by cdelv** [comment]
Im using this simple mesh generated with gmsh. I will test MFEM default meshes and let you know
```
$MeshFormat
2.2 0 8
$EndMeshFormat
$PhysicalNames
5
1 1 "down"
1 2 "top"
1 3 "left"
1 4 "right"
2 5 "Sea"
$EndPhysicalNames
$Nodes
143
1 0 0 0
2 10 0 0
3 10 10 0
4 0 10 0
5 1 0 0
6 2 0 0
7 3 0 0
8 4 0 0
9 5 0 0
10 6 0 0
11 7 0 0
12 8 0 0
13 9 0 0
14 10 1 0
15 10 2 0
16 10 3 0
17 10 4 0
18 10 5 0
19 10 6 0
20 10 7 0
21 10 8 0
22 10 9 0
23 9 10 0
24 8 10 0
25 7 10 0
26 6 10 0
27 5 10 0
28 4 10 0
29 3 10 0
30 2 10 0
31 1 10 0
32 0 9 0
33 0 8 0
34 0 7 0
35 0 6 0
36 0 5 0
37 0 4 0
38 0 3 0
39 0 2 0
40 0 1 0
41 9.133974596215561 5.5 0
42 4.636303417098427 9.059727590364488 0
43 4.639102936910168 0.9500008811816061 0
44 0.8660254037844384 5.5 0
45 9.133974596215563 3.5 0
46 6.50920199901657 9.1114569792819 0
47 6.512758904849833 0.8857995134062442 0
48 0.8660254037844384 3.5 0
49 2.5 9.133974596215563 0
50 9.138693290134388 7.491826982387224 0
51 2.520415328370953 0.8872509776491073 0
52 0.7959225689634009 7.459788032323782 0
53 8.368818016085651 0.9007714138841676 0
54 8.408174380942597 9.094302752559205 0
55 0.9549528281023978 1.5544405458943 0
56 9.138693290134389 2.508173017612776 0
57 8.268735641417594 3.001362169602128 0
58 8.268080267262201 4.00022702826702 0
59 7.402076709282943 3.500264866311523 0
60 7.401971121224573 4.500081982429755 0
61 6.535931760397936 4.000057808123543 0
62 6.535911836214508 5.000023298425547 0
63 5.669880785559132 4.500013517758179 0
64 5.669876523716736 5.500006136030618 0
65 4.803849468480077 5.000003275631463 0
66 4.803848482930974 6.000001568610344 0
67 4.80384919323804 4.00000279889827 0
68 3.937822325664775 4.500000263541651 0
69 6.535901217194109 6.000004905742692 0
70 3.937822468192349 3.500000510406649 0
71 4.803847895731384 3.000000551550815 0
72 3.071796844197701 4.00000012899138 0
73 3.071796807496 5.000000065422169 0
74 3.071796831250595 3.000000106566334 0
75 6.535929434224238 3.000053779072508 0
76 2.205771365940052 4.499999999999996 0
77 2.205771365940052 5.499999999999996 0
78 3.071796776019743 6.000000010903691 0
79 2.205771365940053 6.499999999999996 0
80 3.0717967707737 7.000000001817279 0
81 2.217101144883348 7.481067630677528 0
82 3.032091287815446 8.145899316879582 0
83 3.93782217350893 7.499999999999996 0
84 8.27501329042887 1.991797573767796 0
85 7.401934391357019 5.500018364432998 0
86 7.401926027820381 6.500003878362613 0
87 6.53589923011317 7.000001464017548 0
88 7.401924302717451 7.500000890396692 0
89 6.528529953070857 8.025174079608163 0
90 5.66836161380589 7.504258351074138 0
91 7.426237728682217 8.507269252472398 0
92 5.607707426714185 8.508900528253136 0
93 4.793453728301794 8.001251432412332 0
94 8.268736100291671 6.998638625191088 0
95 8.267980694117673 5.000054562521629 0
96 5.669876198650419 3.500005572999237 0
97 5.669878745306579 2.50000998393709 0
98 4.803847577293372 1.999999999999993 0
99 6.53589838486225 1.999999999999996 0
100 3.948185257032031 2.517949387185086 0
101 3.035403696774028 1.856955464515777 0
102 2.214991334530517 2.510220805667021 0
103 5.627283536167732 1.501439998622024 0
104 7.432402064145548 1.482844822877428 0
105 9.134001692301833 4.500046931798108 0
106 9.134892196512123 6.498410934596385 0
107 8.301008760416142 8.007166732888324 0
108 3.5 9.133974596215563 0
109 1.505932316993227 9.170502595851184 0
110 5.669622440842375 6.500712070912555 0
111 8.26824066771907 5.999521060850785 0
112 0.8660254037844375 4.5 0
113 0.8154625002685538 2.540998849579219 0
114 3.5 0.8660254037844375 0
115 0.8660254037844375 6.5 0
116 3.937822325672607 6.500000263555217 0
117 2.205771365940051 3.499999999999996 0
118 3.937822697710696 5.500000907944089 0
119 7.408342587393573 2.49605386860523 0
120 4.801821794177094 7.001037281094097 0
121 9.165616675679733 1.542034260871268 0
122 9.179204029437228 8.46947073888594 0
123 1.534846009811132 0.7989182588251706 0
124 0.9385434397696786 8.440933898609099 0
125 1.999999999999999 8.38053117003826 0
126 2.001657094824565 1.619767221838738 0
127 3.999999999999999 1.607695154586731 0
128 3.997918925542147 8.39174722319289 0
129 5.514532766447716 0.6739851280024087 0
130 7.468722821728277 9.342605796862701 0
131 1.535898384862246 4.999999999999996 0
132 1.535898384862248 5.999999999999996 0
133 1.535898384862247 6.999999999999995 0
134 1.535898384862243 2.999999999999999 0
135 1.535898384862247 3.999999999999997 0
136 5.525838045186651 9.341610657998245 0
137 7.473425569297291 0.6438309460829368 0
138 0.7320508075688759 9.267949192431125 0
139 0.7320508075688767 0.7320508075688767 0
140 9.267949192431123 9.267949192431123 0
141 9.267949192431123 0.7320508075688767 0
142 1.513536987473252 2.229790360877329 0
143 1.497493107695735 7.752464146329734 0
$EndNodes
$Elements
284
1 1 2 1 1 1 5
2 1 2 1 1 5 6
3 1 2 1 1 6 7
4 1 2 1 1 7 8
5 1 2 1 1 8 9
6 1 2 1 1 9 10
7 1 2 1 1 10 11
8 1 2 1 1 11 12
9 1 2 1 1 12 13
10 1 2 1 1 13 2
11 1 2 4 2 2 14
12 1 2 4 2 14 15
13 1 2 4 2 15 16
14 1 2 4 2 16 17
15 1 2 4 2 17 18
16 1 2 4 2 18 19
17 1 2 4 2 19 20
18 1 2 4 2 20 21
19 1 2 4 2 21 22
20 1 2 4 2 22 3
21 1 2 2 3 3 23
22 1 2 2 3 23 24
23 1 2 2 3 24 25
24 1 2 2 3 25 26
25 1 2 2 3 26 27
26 1 2 2 3 27 28
27 1 2 2 3 28 29
28 1 2 2 3 29 30
29 1 2 2 3 30 31
30 1 2 2 3 31 4
31 1 2 3 4 4 32
32 1 2 3 4 32 33
33 1 2 3 4 33 34
34 1 2 3 4 34 35
35 1 2 3 4 35 36
36 1 2 3 4 36 37
37 1 2 3 4 37 38
38 1 2 3 4 38 39
39 1 2 3 4 39 40
40 1 2 3 4 40 1
41 2 2 5 1 43 103 98
42 2 2 5 1 82 108 49
43 2 2 5 1 51 114 101
44 2 2 5 1 42 93 92
45 2 2 5 1 84 104 53
46 2 2 5 1 91 107 54
47 2 2 5 1 43 129 103
48 2 2 5 1 49 125 82
49 2 2 5 1 101 126 51
50 2 2 5 1 92 136 42
51 2 2 5 1 54 130 91
52 2 2 5 1 89 91 46
53 2 2 5 1 104 137 53
54 2 2 5 1 46 92 89
55 2 2 5 1 47 104 99
56 2 2 5 1 114 127 101
57 2 2 5 1 82 128 108
58 2 2 5 1 99 103 47
59 2 2 5 1 84 119 104
60 2 2 5 1 107 122 54
61 2 2 5 1 53 121 84
62 2 2 5 1 104 119 99
63 2 2 5 1 124 125 109
64 2 2 5 1 82 125 81
65 2 2 5 1 102 126 101
66 2 2 5 1 98 127 43
67 2 2 5 1 89 92 90
68 2 2 5 1 42 128 93
69 2 2 5 1 123 126 55
70 2 2 5 1 92 93 90
71 2 2 5 1 124 143 125
72 2 2 5 1 91 130 46
73 2 2 5 1 83 120 93
74 2 2 5 1 93 120 90
75 2 2 5 1 94 106 50
76 2 2 5 1 50 107 94
77 2 2 5 1 74 100 70
78 2 2 5 1 58 105 95
79 2 2 5 1 71 97 96
80 2 2 5 1 74 101 100
81 2 2 5 1 70 100 71
82 2 2 5 1 71 98 97
83 2 2 5 1 94 111 106
84 2 2 5 1 106 111 41
85 2 2 5 1 45 105 58
86 2 2 5 1 63 96 61
87 2 2 5 1 97 103 99
88 2 2 5 1 94 107 88
89 2 2 5 1 78 116 80
90 2 2 5 1 60 61 59
91 2 2 5 1 60 85 62
92 2 2 5 1 61 96 75
93 2 2 5 1 62 69 64
94 2 2 5 1 64 110 66
95 2 2 5 1 68 70 67
96 2 2 5 1 85 111 86
97 2 2 5 1 88 107 91
98 2 2 5 1 17 105 45
99 2 2 5 1 19 106 41
100 2 2 5 1 28 108 42
101 2 2 5 1 30 109 49
102 2 2 5 1 95 105 41
103 2 2 5 1 41 111 95
104 2 2 5 1 45 58 57
105 2 2 5 1 60 95 85
106 2 2 5 1 62 85 69
107 2 2 5 1 65 118 68
108 2 2 5 1 110 120 66
109 2 2 5 1 70 71 67
110 2 2 5 1 86 87 69
111 2 2 5 1 78 118 116
112 2 2 5 1 74 102 101
113 2 2 5 1 80 82 81
114 2 2 5 1 90 120 110
115 2 2 5 1 71 96 67
116 2 2 5 1 71 100 98
117 2 2 5 1 58 95 60
118 2 2 5 1 62 63 61
119 2 2 5 1 64 65 63
120 2 2 5 1 65 67 63
121 2 2 5 1 66 118 65
122 2 2 5 1 85 86 69
123 2 2 5 1 72 117 74
124 2 2 5 1 95 111 85
125 2 2 5 1 86 88 87
126 2 2 5 1 89 90 87
127 2 2 5 1 7 51 6
128 2 2 5 1 9 43 8
129 2 2 5 1 43 114 8
130 2 2 5 1 11 47 10
131 2 2 5 1 13 53 12
132 2 2 5 1 34 52 33
133 2 2 5 1 36 44 35
134 2 2 5 1 44 115 35
135 2 2 5 1 38 48 37
136 2 2 5 1 40 55 39
137 2 2 5 1 68 73 72
138 2 2 5 1 73 77 76
139 2 2 5 1 74 117 102
140 2 2 5 1 99 119 75
141 2 2 5 1 80 81 79
142 2 2 5 1 68 72 70
143 2 2 5 1 72 74 70
144 2 2 5 1 73 76 72
145 2 2 5 1 76 117 72
146 2 2 5 1 73 118 78
147 2 2 5 1 78 80 79
148 2 2 5 1 80 83 82
149 2 2 5 1 7 114 51
150 2 2 5 1 34 115 52
151 2 2 5 1 36 112 44
152 2 2 5 1 38 113 48
153 2 2 5 1 55 113 39
154 2 2 5 1 61 75 59
155 2 2 5 1 88 91 89
156 2 2 5 1 8 114 7
157 2 2 5 1 35 115 34
158 2 2 5 1 37 112 36
159 2 2 5 1 48 112 37
160 2 2 5 1 39 113 38
161 2 2 5 1 58 59 57
162 2 2 5 1 57 119 84
163 2 2 5 1 65 68 67
164 2 2 5 1 73 78 77
165 2 2 5 1 58 60 59
166 2 2 5 1 75 119 59
167 2 2 5 1 60 62 61
168 2 2 5 1 62 64 63
169 2 2 5 1 67 96 63
170 2 2 5 1 64 66 65
171 2 2 5 1 69 110 64
172 2 2 5 1 116 118 66
173 2 2 5 1 97 99 75
174 2 2 5 1 78 79 77
175 2 2 5 1 80 116 83
176 2 2 5 1 86 111 94
177 2 2 5 1 88 89 87
178 2 2 5 1 98 103 97
179 2 2 5 1 59 119 57
180 2 2 5 1 68 118 73
181 2 2 5 1 96 97 75
182 2 2 5 1 90 110 87
183 2 2 5 1 16 56 15
184 2 2 5 1 17 45 16
185 2 2 5 1 45 56 16
186 2 2 5 1 18 105 17
187 2 2 5 1 19 41 18
188 2 2 5 1 41 105 18
189 2 2 5 1 20 106 19
190 2 2 5 1 21 50 20
191 2 2 5 1 50 106 20
192 2 2 5 1 24 54 23
193 2 2 5 1 26 46 25
194 2 2 5 1 28 42 27
195 2 2 5 1 29 108 28
196 2 2 5 1 30 49 29
197 2 2 5 1 49 108 29
198 2 2 5 1 31 109 30
199 2 2 5 1 45 57 56
200 2 2 5 1 57 84 56
201 2 2 5 1 66 120 116
202 2 2 5 1 87 110 69
203 2 2 5 1 86 94 88
204 2 2 5 1 116 120 83
205 2 2 5 1 126 142 55
206 2 2 5 1 50 122 107
207 2 2 5 1 15 121 14
208 2 2 5 1 6 123 5
209 2 2 5 1 33 124 32
210 2 2 5 1 22 122 21
211 2 2 5 1 84 121 56
212 2 2 5 1 103 129 47
213 2 2 5 1 46 136 92
214 2 2 5 1 101 127 100
215 2 2 5 1 83 128 82
216 2 2 5 1 47 137 104
217 2 2 5 1 51 123 6
218 2 2 5 1 52 124 33
219 2 2 5 1 56 121 15
220 2 2 5 1 21 122 50
221 2 2 5 1 47 129 10
222 2 2 5 1 9 129 43
223 2 2 5 1 46 130 25
224 2 2 5 1 24 130 54
225 2 2 5 1 26 136 46
226 2 2 5 1 42 136 27
227 2 2 5 1 53 137 12
228 2 2 5 1 11 137 47
229 2 2 5 1 5 139 1
230 2 2 5 1 1 139 40
231 2 2 5 1 2 141 13
232 2 2 5 1 14 141 2
233 2 2 5 1 3 140 22
234 2 2 5 1 23 140 3
235 2 2 5 1 32 138 4
236 2 2 5 1 4 138 31
237 2 2 5 1 109 125 49
238 2 2 5 1 93 128 83
239 2 2 5 1 108 128 42
240 2 2 5 1 100 127 98
241 2 2 5 1 43 127 114
242 2 2 5 1 51 126 123
243 2 2 5 1 55 142 113
244 2 2 5 1 10 129 9
245 2 2 5 1 131 132 44
246 2 2 5 1 115 133 52
247 2 2 5 1 25 130 24
248 2 2 5 1 112 131 44
249 2 2 5 1 44 132 115
250 2 2 5 1 48 135 112
251 2 2 5 1 117 134 102
252 2 2 5 1 132 133 115
253 2 2 5 1 77 132 131
254 2 2 5 1 112 135 131
255 2 2 5 1 113 134 48
256 2 2 5 1 77 131 76
257 2 2 5 1 131 135 76
258 2 2 5 1 76 135 117
259 2 2 5 1 79 132 77
260 2 2 5 1 81 133 79
261 2 2 5 1 134 135 48
262 2 2 5 1 79 133 132
263 2 2 5 1 117 135 134
264 2 2 5 1 27 136 26
265 2 2 5 1 52 143 124
266 2 2 5 1 124 138 32
267 2 2 5 1 123 139 5
268 2 2 5 1 22 140 122
269 2 2 5 1 121 141 14
270 2 2 5 1 12 137 11
271 2 2 5 1 13 141 53
272 2 2 5 1 40 139 55
273 2 2 5 1 54 140 23
274 2 2 5 1 31 138 109
275 2 2 5 1 109 138 124
276 2 2 5 1 53 141 121
277 2 2 5 1 122 140 54
278 2 2 5 1 55 139 123
279 2 2 5 1 125 143 81
280 2 2 5 1 102 142 126
281 2 2 5 1 134 142 102
282 2 2 5 1 113 142 134
283 2 2 5 1 133 143 52
284 2 2 5 1 81 143 133
$EndElements
```

**Comment by cdelv** [comment]
@jandrej I used MFEM star mesh and the same thing happened.

**Comment by jandrej** [comment]
Here is the _peak_ memory usage for `100 loops`
```
	Command being timed: "./examples/gh2413 --mesh ../data/inline-quad.mesh"
	User time (seconds): 3.58
	System time (seconds): 1.09
	Percent of CPU this job got: 99%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:04.68
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 50276
	Average resident set size (kbytes): 0
```
and for 1000 loops
```
	Command being timed: "./examples/gh2413 --mesh ../data/inline-quad.mesh"
	User time (seconds): 34.15
	System time (seconds): 10.64
	Percent of CPU this job got: 99%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:44.84
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 50276
	Average resident set size (kbytes): 0
```
The value of `Maximum resident set size` indicates the peak memory usage.

**Comment by cdelv** [comment]
@jandrej 
Running the same tes I have for 100 loops 
```
	Command being timed: "mpirun -np 4 ./a.out"
	User time (seconds): 19.00
	System time (seconds): 0.74
	Percent of CPU this job got: 373%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:05.28
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 360760
```
and for 1000
```
Command being timed: "mpirun -np 4 ./a.out"
	User time (seconds): 241.10
	System time (seconds): 6.51
	Percent of CPU this job got: 397%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 1:02.26
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 3289836
```

**Comment by jandrej** [comment]
Then we narrowed the problem down to the mesh it seems. I am not familiar with gmsh meshes causing these types of issues. They are used by a variety of users.

**Comment by cdelv** [comment]
@jandrej It seems to be something of the parallel rutin. I ran the test again not using mpirun and these are the results
for 100 loop
```
	Command being timed: "./a.out"
	User time (seconds): 4.41
	System time (seconds): 0.02
	Percent of CPU this job got: 92%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:04.79
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 50700
```
for 1000 loops
```
	Command being timed: "./a.out"
	User time (seconds): 45.04
	System time (seconds): 0.09
	Percent of CPU this job got: 99%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:45.50
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 50744
```
Also, we have used this same mesh for multiple programs before and dint have these problems.

**Comment by jandrej** [comment]
Duplicate of #2368. Closing.

**Comment by psocratis** [comment]
@cdelv, can you test using #2420 ?

**Comment by cdelv** [comment]
@psocratis help me, please. I'm a bit clueless about what should I do. The change is already in the master branch? so, I reinstall MFEM? after that, how do I approve?

**Comment by psocratis** [comment]
> @psocratis help me, please. I'm a bit clueless about what should I do. The change is already in the master branch? so, I reinstall MFEM? after that, how do I approve?

@cdelv, You can repull `master` and then switch to the branch `slu-destroy-fix`. Then recompile mfem and try your example against the new built. 

Don't worry about approving, just let us know if if the issue is resolved.

Thanks, 

Socratis

**Comment by cdelv** [comment]
@psocratis ok, give me 30 min

**Comment by cdelv** [comment]
@psocratis The fix worked
